book/src/dev/rfcs/drafts/xxxx-async-rust.md

```diff
@@ -0,0 +1,438 @@
+  - in general, we should move as much work into futures as possible, unless the design requires sequential `call`s
+- larger `Buffer`/`Batch` bounds
+
+### Buffered Services
+[buffered-services]: #buffered-services
+
+A service should be provided wrapped in a `Buffer` if:
+* it is a complex service
+* it has multiple callers, or
+* it has a single caller than calls it multiple times concurrently.
+
+Services might also have other reasons for using a `Buffer`. These reasons should be documented.
+
+#### Choosing Buffer Bounds
+[choosing-buffer-bounds]: #choosing-buffer-bounds
+
+Zebra's `Buffer` bounds should be set to the maximum number of concurrent requests, plus 1:
+> it's advisable to set bound to be at least the maximum number of concurrent requests the `Buffer` will see
+https://docs.rs/tower/0.4.3/tower/buffer/struct.Buffer.html#method.new
+
+The extra slot protects us from future changes that add an extra caller, or extra concurrency.
+
+As a general rule, Zebra `Buffer`s should all have at least 3 slots (2 + 1), because most Zebra services can
+be called concurrently by:
+* the sync service, and
+* the inbound service.
+
+Services might also have other reasons for a larger bound. These reasons should be documented.
+
+We should limit `Buffer` lengths for services whose requests or responses contain `Block`s (or other large
+data items, such as `Transaction` vectors). A long `Buffer` full of `Block`s can significantly increase memory
+usage.
+
+Long `Buffer`s can also increase request latency. Latency isn't a concern for Zebra's core use case as a node
+software, but it might be an issue if wallets, exchanges, or block explorers want to use Zebra.
+
+## Awaiting Multiple Futures
+[awaiting-multiple-futures]: #awaiting-multiple-futures
+
+When awaiting multiple futures, Zebra can use biased or unbiased selection.
+
+Typically, we prefer unbiased selection, so that if multiple futures are ready,
+they each have a chance of completing. But if one of the futures needs to take
+priority (for example, cancellation), you might want to use biased selection.
+
+### Unbiased Selection
+[unbiased-selection]: #unbiased-selection
+
+The [`futures::select!`](https://docs.rs/futures/0.3.13/futures/macro.select.html) and
+[`tokio::select!`](https://docs.rs/tokio/0.3.6/tokio/macro.select.html) macros select
+ready arguments at random.
+
+Also consdier the `FuturesUnordered` stream for unbiased selection of a large number
+of futures. However, this macro and stream require mapping all arguments to the same
+type.
+
+Consider mapping the returned type to a custom enum with module-specific names.
+
+### Biased Selection
+[biased-selection]: #biased-selection
+
+The [`futures::select`](https://docs.rs/futures/0.3.13/futures/future/fn.select.html)
+is biased towards its first argument. If the first argument is always ready, the second
+argument will never be returned. (This behavior is not documented or guaranteed.) This
```

Comment:
😧

Reply:
Yeah we might want to submit futures-rs and tokio docs PRs for the select bias.